File: ports/csharp/src/core/tools/BcsCommandLineParser.cs
using System;
using System.Collections.Generic;
using BCS.Core.Global;

namespace BCS.Core.Tools {
    public class BcsCommandLineOption {
        public BcsString Name { get; set; }
        public BcsString Description { get; set; }
        public BcsString ValueName { get; set; }

        public BcsCommandLineOption(BcsString name, BcsString description, BcsString valueName) {
            Name = name;
            Description = description;
            ValueName = valueName;
        }
    }

    public class BcsCommandLineParser {
        private Dictionary<string, BcsCommandLineOption> options;
        private Dictionary<string, BcsString> values;

        public BcsCommandLineParser() {
            options = new Dictionary<string, BcsCommandLineOption>();
            values = new Dictionary<string, BcsString>();
        }

        public void AddOption(BcsCommandLineOption opt) {
            options[opt.Name.ToString()] = opt;
        }

        public void Process(string[] args) {
            for (int i = 0; i < args.Length; i++) {
                string arg = args[i];
                if (arg.Length > 2 && arg.StartsWith("--")) {
                    string flag = arg.Substring(2);
                    if (options.ContainsKey(flag)) {
                        values[flag] = new BcsString("true");
                    }
                }
            }
        }

        public bool IsSet(BcsString name) {
            return values.ContainsKey(name.ToString());
        }

        public BcsString Value(BcsString name) {
            if (values.TryGetValue(name.ToString(), out var val)) {
                return val;
            }
            return null;
        }

        public void ShowHelp() {
            Console.WriteLine("Usage:");
            foreach (var kvp in options) {
                Console.WriteLine($"  --{kvp.Key}\t{kvp.Value.Description}");
            }
            Environment.Exit(0);
        }
    }
}

File: ports/rust/src/core/tools/bcscommandlineparser.rs
use crate::core::global::bcsstring::BcsStringWrapper;
use std::collections::HashMap;

pub struct BcsCommandLineOption {
    pub name: BcsStringWrapper,
    pub description: BcsStringWrapper,
    pub value_name: BcsStringWrapper,
}

pub struct BcsCommandLineParser {
    options: HashMap<String, BcsCommandLineOption>,
    values: HashMap<String, BcsStringWrapper>,
}

impl BcsCommandLineParser {
    pub fn new() -> Self {
        BcsCommandLineParser {
            options: HashMap::new(),
            values: HashMap::new(),
        }
    }

    pub fn add_option(&mut self, opt: BcsCommandLineOption) {
        self.options.insert(opt.name.to_string(), opt);
    }

    pub fn process(&mut self, args: Vec<String>) {
        for arg in args {
            if arg.len() > 2 && arg.starts_with("--") {
                let flag = &arg[2..];
                if self.options.contains_key(flag) {
                    self.values.insert(flag.to_string(), BcsStringWrapper::new("true"));
                }
            }
        }
    }

    pub fn is_set(&self, name: &BcsStringWrapper) -> bool {
        self.values.contains_key(&name.to_string())
    }

    pub fn value(&self, name: &BcsStringWrapper) -> Option<&BcsStringWrapper> {
        self.values.get(&name.to_string())
    }

    pub fn show_help(&self) {
        println!("Usage:");
        for (name, opt) in &self.options {
            println!("  --{}\t{}", name, opt.description.to_string());
        }
        std::process::exit(0);
    }
}
